test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc

```diff
@@ -187,6 +195,69 @@ TEST(DefaultCertValidatorTest, NoSanInCert) {
   EXPECT_FALSE(DefaultCertValidator::matchSubjectAltName(cert.get(), subject_alt_name_matchers));
 }
 
+TEST(DefaultCertValidatorTest, WithVerifyDepth) {
+
+  Stats::TestUtil::TestStore test_store;
+  SslStats stats = generateSslStats(test_store);
+  envoy::config::core::v3::TypedExtensionConfig typed_conf;
+  std::vector<envoy::extensions::transport_sockets::tls::v3::SubjectAltNameMatcher> san_matchers{};
+
+  bssl::UniquePtr<STACK_OF(X509)> cert_chain = readCertChainFromFile(TestEnvironment::substitute(
+      "{{ test_rundir "
+      "}}/test/extensions/transport_sockets/tls/test_data/test_long_cert_chain.pem"));
+  bssl::UniquePtr<X509> cert = readCertFromFile(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/test_random_cert.pem"));
+  bssl::UniquePtr<X509> ca_cert = readCertFromFile(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/ca_cert.pem"));
+
+  // Create the default validator object.
+  // Config includes ca_cert and the verify-depth
+  // Set verify depth < 3, so verification fails. ( There are 3 intermediate certs )
+
+  std::string ca_cert_str(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/ca_cert.pem"));
+  TestCertificateValidationContextConfigPtr test_config =
+      std::make_unique<TestCertificateValidationContextConfig>(typed_conf, false, san_matchers,
+                                                               ca_cert_str, 2);
+  auto default_validator =
+      std::make_unique<Extensions::TransportSockets::Tls::DefaultCertValidator>(
+          test_config.get(), stats, Event::GlobalTimeSystem().timeSystem());
+
+  STACK_OF(X509)* intermediates = cert_chain.get();
+  SSLContextPtr ssl_ctx = SSL_CTX_new(TLS_method());
+  X509StoreContextPtr store_ctx = X509_STORE_CTX_new();
+
+  X509_STORE* storep = SSL_CTX_get_cert_store(ssl_ctx.get());
+  X509_STORE_add_cert(storep, ca_cert.get());
+  EXPECT_TRUE(X509_STORE_CTX_init(store_ctx.get(), storep, cert.get(), intermediates));
+
+  default_validator->addClientValidationContext(ssl_ctx.get(), false);
+  X509_VERIFY_PARAM_set1(X509_STORE_CTX_get0_param(store_ctx.get()),
+                         SSL_CTX_get0_param(ssl_ctx.get()));
+
+  EXPECT_EQ(X509_verify_cert(store_ctx.get()), 0);
```

Comment:
Can the test validate the behavior using the validator interface, instead of reaching into the guts of the SSL library? Something more like:

```
EXPECT_EQ(default_validator->verifyCertificate(cert.get(), /*verify_san_list=*/{},
                                                 invalid_san_matchers),
            Envoy::Ssl::ClientValidationStatus::Failed);
```

Reply:
verifyCertificate() requires valid (non-null) san_list/subject_alt_name_matchers/verify_certificate_hash_list which would be all orthagonal to this logic. Ideally, i wanted to use doVerifyCertChain(). This requires setting the ca_cert in trusted store done in initializeSslContexts. The SSL library calls in the test essentially replace that functionality.. Create a store_ctx and set the ca_cert. Without this, I'll have to rewrite the defaultValidator test to handle all the different configurations (sth like what the spiffee validator test does today), which looked way too extensive!

Reply:
As we discussed offline, i'll leave the unit-test as-is and add an integration test as well.